src/monty/io.py

```diff
@@ -14,40 +14,89 @@
 
-def zopen(filename: Union[str, Path], *args, **kwargs) -> IO:
+def zopen(
+    filename: Union[str, Path],
+    /,
+    mode: str | None = None,
+    **kwargs: Any,
+) -> IO:  # TODO: incomplete return type for compressed files
     """
-    This function wraps around the bz2, gzip, lzma, xz and standard Python's open
-    function to deal intelligently with bzipped, gzipped or standard text
-    files.
+    This function wraps around `[bz2/gzip/lzma].open` and `open`
+    to deal intelligently with compressed or uncompressed files.
+    Supports context manager:
+        `with zopen(filename, mode="rt", ...)`.
+
+    Important Notes:
+        - Default `mode` should not be used, and would not be allow
+            in future versions.
+        - Always explicitly specify binary/text in `mode`, i.e.
+            always pass `t` or `b` in `mode`.
+        - When using text mode, always provide an explicit encoding.
 
     Args:
-        filename (str/Path): filename or pathlib.Path.
-        *args: Standard args for Python open(..). E.g., 'r' for read, 'w' for
-            write.
-        **kwargs: Standard kwargs for Python open(..).
+        filename (str | Path): Filename.
+        mode (str): The mode in which the file is opened (
+            e.g., "r" for read, "w" for write.), you MUST
+            explicitly specify "b" for binary or "t" for text.
+        **kwargs: Additional keyword arguments to pass to `open`.
 
     Returns:
-        File-like object. Supports with context.
+        TextIO | BinaryIO
     """
-    if filename is not None and isinstance(filename, Path):
-        filename = str(filename)
-
-    _name, ext = os.path.splitext(filename)
-    ext = ext.upper()
-    if ext == ".BZ2":
-        return bz2.open(filename, *args, **kwargs)
-    if ext in {".GZ", ".Z"}:
-        return gzip.open(filename, *args, **kwargs)
-    if ext in {".XZ", ".LZMA"}:
-        return lzma.open(filename, *args, **kwargs)
-    return open(filename, *args, **kwargs)
+    _deadline = "2025-06-01"
+
+    # Warn against default `mode`
+    # TODO: remove default value of `mode` to force user to give one after deadline
+    if mode is None:
+        warnings.warn(
+            "We strongly discourage using a default `mode`, it would be"
+            f"set to `rt` now but would not be allowed after {_deadline}",
+            FutureWarning,
+            stacklevel=2,
+        )  # TODO: unit test
+        mode = "rt"
+
+    # Warn against implicit text/binary `mode`
+    # TODO: replace warning with exception after deadline
+    elif not ("b" in mode or "t" in mode):
+        warnings.warn(
+            "We strongly discourage using implicit binary/text `mode`, "
+            f"and this would not be allowed after {_deadline}. "
+            "I.e. you should pass t/b in `mode`, we would assume text mode for now",
+            FutureWarning,
+            stacklevel=2,
+        )  # TODO: unit test
+        mode += "t"  # assume text mode if not specified
```

Comment:
I would assume `mode` to be text here instead of leave it as is for two reasons:
- For `pymatgen`, [most (if not all) implicit binary/text usage seem to assume text mode](https://github.com/search?q=repo%3Amaterialsproject%2Fpymatgen%20with%20zopen&type=code), i.e. apply string-specific operations on the stream instead of bytes
- Facilitate `encoding` check (otherwise cannot use `if "t" in encoding` to determine if text mode is expected)